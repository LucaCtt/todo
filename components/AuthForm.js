import React, { useState } from "react";
import { View } from "react-native";
import { Text, Spinner } from "@ui-kitten/components";

import FormInput from "../components/FormInput";
import FormButton from "../components/FormButton";

const LoadingIndicator = (props) => (
  <View {...props}>
    <Spinner />
  </View>
);

export default AuthForm = ({ submitText, onSubmit, onSuccess }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);

  const submit = async () => {
    try {
      setLoading(true);
      await onSubmit(email, password);
      setLoading(false);
      onSuccess(email);
    } catch (err) {
      setError(err.message);
      setLoading(false);
    }
  };

  return (
    <>
      <FormInput
        autoFocus
        keyboardType="email-address"
        placeholder="Email"
        value={email}
        onChangeText={(e) => setEmail(e)}
      />
      <FormInput
        secureTextEntry
        placeholder="Password"
        value={password}
        onChangeText={(p) => setPassword(p)}
      />
      <FormButton
        disabled={email == "" || password == "" || loading}
        onPress={submit}
        accessoryLeft={loading ? LoadingIndicator : null}
      >
        <Text>{submitText}</Text>
      </FormButton>
      {error !== "" && <Text status="danger">{error}</Text>}
    </>
  );
};
