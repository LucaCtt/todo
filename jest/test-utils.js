import React from "react";
import { render } from "@testing-library/react-native";
import * as eva from "@eva-design/eva";
import { EvaIconsPack } from "@ui-kitten/eva-icons";
import { ApplicationProvider, IconRegistry } from "@ui-kitten/components";

const TestProvider = ({ children }) => (
  <>
    <IconRegistry icons={EvaIconsPack} />
    <ApplicationProvider mapping={eva.mapping} theme={eva.light}>
      {children}
    </ApplicationProvider>
  </>
);

const customRender = (ui, options) =>
  render(ui, { wrapper: TestProvider, ...options });

// re-export everything
export * from "@testing-library/react-native";

// override render method
export { customRender as render };
