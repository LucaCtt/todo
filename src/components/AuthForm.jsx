import React, { useState, memo } from "react";
import { View } from "react-native";
import { Text } from "@ui-kitten/components";

import FormInput from "./FormInput";
import SubmitButton from "./SubmitButton";

/**
 * A form with inputs for email and password, and a submit button.
 *
 * @property {string} submitText - The text shown on the submit button.
 * @property {() => void} onSubmit - The function to be run on submit. Can be async: in that case the result
 *                                   will be awaited. Thrown errors will be shown in the form.
 */
const AuthForm = ({ submitText, onSubmit, ...props }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);

  const submit = async () => {
    setLoading(true);
    try {
      await onSubmit(email, password);
    } catch (err) {
      setError(err.message);
      setLoading(false);
    }
  };

  return (
    <View {...props}>
      <FormInput
        autoFocus
        keyboardType="email-address"
        placeholder="Email"
        value={email}
        onChangeText={(e) => setEmail(e)}
      />
      <FormInput
        secureTextEntry
        placeholder="Password"
        value={password}
        onChangeText={(p) => setPassword(p)}
      />
      <SubmitButton
        disabled={email === "" || password === "" || loading}
        onPress={submit}
        isLoading={loading}
      >
        <Text>{submitText}</Text>
      </SubmitButton>
      {error !== "" && <Text status="danger">{error}</Text>}
    </View>
  );
};

export default memo(AuthForm);
