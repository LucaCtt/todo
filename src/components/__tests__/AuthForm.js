import React from "react";
import { render, fireEvent } from "test-utils";

import AuthForm from "../AuthForm";

describe("<AuthForm />", () => {
  test("matches snapshot", () => {
    const { baseElement } = render(
      <AuthForm submitText="Test" onSubmit={() => {}} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  test("onSubmit called with email and password", () => {
    const onSubmit = jest.fn();

    const { getByPlaceholderText, getByText } = render(
      <AuthForm submitText="Submit" onSubmit={onSubmit} />
    );

    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "12345678Aa@");
    fireEvent.press(getByText("Submit"));

    expect(onSubmit.mock.calls[0]).toEqual(["test@example.com", "12345678Aa@"]);
  });

  /*
  test("onSubmit error is shown", () => {
    const onSubmit = () => {
      throw new Error("test");
    };

    const { getByText, baseElement } = render(
      <AuthForm submitText="Submit" onSubmit={onSubmit} />
    );

    fireEvent.press(getByText("Submit"));

    expect(baseElement).toContainElement(getByText("test"));
  });*/
});
