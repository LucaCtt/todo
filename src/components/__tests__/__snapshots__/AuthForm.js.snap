// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthForm /> matches snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginVertical": 6,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#F7F9FC",
                  "borderColor": "#E4E9F2",
                  "borderRadius": 4,
                  "borderWidth": 1,
                  "minHeight": 48,
                  "paddingHorizontal": 8,
                  "paddingVertical": 11,
                },
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              appearance="default"
              autoFocus={true}
              editable={true}
              keyboardType="email-address"
              placeholder="Email"
              placeholderTextColor="#8F9BB3"
              rejectResponderTermination={true}
              size="large"
              status="basic"
              style={
                Array [
                  Object {
                    "color": "#222B45",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "normal",
                    "marginHorizontal": 8,
                  },
                  Object {
                    "flexBasis": "auto",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                  null,
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 4,
                },
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "marginVertical": 6,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#F7F9FC",
                  "borderColor": "#E4E9F2",
                  "borderRadius": 4,
                  "borderWidth": 1,
                  "minHeight": 48,
                  "paddingHorizontal": 8,
                  "paddingVertical": 11,
                },
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              appearance="default"
              editable={true}
              placeholder="Password"
              placeholderTextColor="#8F9BB3"
              rejectResponderTermination={true}
              secureTextEntry={true}
              size="large"
              status="basic"
              style={
                Array [
                  Object {
                    "color": "#222B45",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "normal",
                    "marginHorizontal": 8,
                  },
                  Object {
                    "flexBasis": "auto",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                  null,
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 4,
                },
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
              ]
            }
          />
        </View>
        <TouchableOpacity
          activeOpacity={1}
          appearance="filled"
          disabled={true}
          size="medium"
          status="primary"
          style={
            Array [
              undefined,
              Array [
                Object {
                  "backgroundColor": "rgba(143, 155, 179, 0.24)",
                  "borderColor": "rgba(143, 155, 179, 0.24)",
                  "borderRadius": 4,
                  "borderWidth": 1,
                  "minHeight": 40,
                  "minWidth": 40,
                  "paddingHorizontal": 10,
                  "paddingVertical": 12,
                },
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                Array [
                  Object {
                    "marginVertical": 10,
                  },
                  undefined,
                ],
              ],
            ]
          }
        >
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "color": "rgba(143, 155, 179, 0.48)",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginHorizontal": 10,
                },
              ]
            }
          >
            <Text
              appearance="default"
              category="p1"
              style={
                Array [
                  Object {
                    "color": "#222B45",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "400",
                  },
                  undefined,
                ]
              }
            >
              Test
            </Text>
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>
</View>
`;
